packages/react-instantsearch-hooks-server/src/getServerState.tsx

```diff
@@ -16,6 +19,18 @@ import type {
  */
 export function getServerState(
   children: ReactNode
+): Promise<InstantSearchServerState> {
+  return renderAndFetch(children, false);
+}
+
+/**
+ * Implementation of getServerState
+ * @param children the application
+ * @param refetching pass true if this is the second time being called
+ */
+function renderAndFetch(
+  children: ReactNode,
+  refetching: boolean
```

Comment:
I find that flipping the naming makes the code and conditions easier to read.

```suggestion
  shouldRefetch: boolean
```

Then:

```ts
  let hasDynamicWidgets = false;
  if (shouldRefetch) {
    // ...
  }

  if (!shouldRefetch || !hasDynamicWidgets) {
    return serverState;
  }

  return renderAndFetch(
    <InstantSearchSSRProvider {...serverState}>
      {children}
    </InstantSearchSSRProvider>,
    false
  );
```

Reply:
what do you think about `canRefetch` with your logic?

Reply:
"can" doesn't reflect what the argument is doing, is it?

Reply:
+1 on "should"; it's a flag, not a test.

Reply:
I'm not sure if "shouldRefetch" being true for the initial call, as it should only refetch if that flag is set + it is actually using dynamic widgets; it won't refetch in other cases

Reply:
I think that the code is not correctly coupled. We don't need this flag.

Here's a proposal, without the concept of "is refetching". I renamed `hasDynamicWidgets` to a generic `shouldRefetch` though, but there's only one variable to track this value.

```tsx
export function getServerState(
  children: ReactNode
): Promise<InstantSearchServerState> {
  const searchRef: { current: InstantSearch | undefined } = {
    current: undefined,
  };

  const notifyServer: InstantSearchServerContextApi['notifyServer'] = ({
    search,
  }) => {
    searchRef.current = search;
  };

  // eslint-disable-next-line no-shadow
  function execute(children: ReactNode) {
    ReactDOM.renderToString(
      <InstantSearchServerContext.Provider value={{ notifyServer }}>
        {children}
      </InstantSearchServerContext.Provider>
    );

    // We wait for the component to mount so that `notifyServer()` is called.
    return new Promise((resolve) => setTimeout(resolve, 0))
      .then(() => {
        // If `notifyServer()` is not called by then, it means that <InstantSearch>
        // wasn't within the `children`.
        // We decide to go with a strict behavior in that case; throwing. If users have
        // some routes that don't mount the <InstantSearch> component, they would need
        // to try/catch the `getServerState()` call.
        // If this behavior turns out to be too strict for many users, we can decide
        // to warn instead of throwing.
        if (!searchRef.current) {
          throw new Error(
            "Unable to retrieve InstantSearch's server state in `getServerState()`. Did you mount the <InstantSearch> component?"
          );
        }

        return waitForResults(searchRef.current);
      })
      .then(() => {
        const initialResults = getInitialResults(searchRef.current!.mainIndex);

        return {
          initialResults,
        };
      });
  }

  return execute(children).then((serverState) => {
    let shouldRefetch = false;

    // <DynamicWidgets> requires another query to retrieve the dynamic widgets
    // to render.
    walkIndex(searchRef.current!.mainIndex, (index) => {
      shouldRefetch =
        shouldRefetch ||
        index
          .getWidgets()
          .some((widget) => widget.$$type === 'ais.dynamicWidgets');
    });

    if (shouldRefetch) {
      return execute(
        <InstantSearchSSRProvider {...serverState}>
          {children}
        </InstantSearchSSRProvider>
      );
    }

    return serverState;
  });
}
```

The code is more linear, it follows the data flow. It's also less error prone because there's less back and forth and we don't keep track of a state variable.